### ISSUE TITLE
GRIT executed for the first time: it works. RRWP costs 1.2 GB/forward, atom_rrwp's docstring is wrong, pipe.py is unreachable

### ISSUE THREAD

**Issue by kuanlinhuang** [Huang-lab/TRACE-Kin#13]
`12b958c` and `e5f56f5` both say "this has not been executed - verification is py_compile plus a stdlib check". That is an honest note and the right thing to write, so I ran them. Results, in the order they matter.

### GRIT runs, forward and backward

At the `config_v7_grit.json` operating point (43 in-channels, d_model 512, grit_dim 64, 10 layers, 8 heads, ksteps 21), on a 4-molecule batch of rings:

```
FORWARD OK   atom_h (26, 512)   mol_pool (4, 512)   p = None
BACKWARD OK  params with grad 296/300   non-finite grads: none
```

No shape errors, no NaNs, and the `(atom_h, mol_pool, None)` contract matches what `net_v7.py:695` unpacks. The searchsorted bond-fold, the per-head `VeRow` contraction, and the block-diagonal batching all behave.

The 4 parameters without gradient are `layers.9.out_e` and `layers.9.bn_e` - the last layer's pair update, whose output nothing consumes. The reference has the same dead tail for graph-level tasks, so this is faithful rather than wrong; it is ~4k parameters of wasted compute per step, not a bug.

I could not run it inside `net_v7` (rdkit/torch_geometric assembly), so this is the encoder in isolation.

### `atom_rrwp` costs 3x the PNA path's positional encoding, and it is recomputed every forward

`atom_rrwp` materialises `(N, N, steps)` where N is the **batch** atom count, not per-molecule. Measured peak RSS, ring molecules of 18 atoms, against `atom_rwse` at the PNA arm's `steps=16`:

| atoms/batch | `atom_rwse(16)` | `atom_rrwp(21)` | time |
|---|---|---|---|
| 180 | +2 MB | +9 MB | 0.00 s |
| 720 | +39 MB | +137 MB | 0.04 s |
| 1440 | +153 MB | +549 MB | 0.21 s |
| 2304 | +395 MB | **+1215 MB** | 0.71 s |

(over a 258 MB baseline; theory says `3 x N^2 x steps x 4 B` = 1.34 GB at N=2304, which matches.)

At ~2300 atoms per batch that is 1.2 GB of transient allocation for the positional encoding alone, before weights or activations, quadratic in batch size. It is not retained by autograd - `edge_index` is integral, so nothing requires grad - but it is a real peak, and on a B200 it competes with the protein stream's dense batching.

The reference avoids this differently rather than better: GRIT computes RRWP once as a dataset `pre_transform` (`add_full_rrwp`), so it is paid at load and cached per graph. Here it is recomputed identically every forward, every epoch, for molecules that never change. `atom_rwse` has the same recompute property in the PNA arm, so this is not new - GRIT just makes it 3x more expensive and adds the pair tensor.

Worth doing before the GRIT arm trains at scale, in rough order of payoff: cache per molecule at featurisation time; or drop pairs by hop distance rather than by nonzero-at-any-k (at ksteps 21 nearly every within-molecule pair survives, so the pruning that keeps P small is doing almost nothing at this walk length).

### `atom_rrwp`'s docstring describes a different return than the code

```
node_pe    (N, steps)   diag((D^-1 A)^k) for k = 2..steps+1  -- identical to
                        atom_rwse, so the two arms stay comparable.
```

Actual: `(N, steps-2)`, covering k = 2..steps-1. Verified:

```
steps=  5   atom_rrwp node_pe (6, 3)    atom_rwse (6, 5)   first 3 cols equal: True
steps=  8   atom_rrwp node_pe (6, 6)    atom_rwse (6, 8)   first 6 cols equal: True
steps= 21   atom_rrwp node_pe (6, 19)   atom_rwse (6, 21)  first 19 cols equal: True
```

The values agree, so the "diag(RRWP) reproduces atom_rwse" check in the commit message holds - it is the shape and the k-range that are misstated. The inline comment two lines down says `(N, steps-2)` correctly, so this is the docstring only.

The "so the two arms stay comparable" clause is worth a second look though: as shipped, GRIT's node PE covers k=2..20 (19 channels, from `mol_grit_steps: 21`) and the PNA arm's covers k=2..17 (16 channels, from `mol_pe_steps: 16`). Same definition, different walk length. Deliberate is fine - `4710120` argues clearly for keeping GRIT at its published operating point rather than forcing it to the PNA one - but "identical" currently overstates it.

### `pipe.py` is unreachable

404 lines, and nothing in the repo refers to it:

```bash
$ grep -rn "pipe" --exclude-dir=.git . | grep -v "^./models/trace_kin/pipe.py" | grep -vi "pipeline\|pipefail"
(nothing)
```

`e5f56f5` says so ("NOT YET: integration"), so this is a note rather than a complaint - but it is on `main`, it imports at package scope, and the `RephineLayer`/`PIPE` classes have no caller to constrain their signatures. Either a tracking issue or a branch would keep `main` honest about what is wired.

### Reproducing

```bash
python - <<'PY'
import torch
from models.trace_kin.grit import GRITLigandEncoder
enc = GRITLigandEncoder(mol_in_channels=43, d_model=512, n_layers=10, heads=8,
                        grit_dim=64, pe_steps=21, pe_raw_norm="batch")
# ... 4-molecule ring batch, forward + backward
PY
```

**Comment by AtlasMaroc** [comment]
Thank you Kuan,

|> 2. atom_rrwp costs 3x the PNA path's positional encoding, and it is recomputed every forward:

for now I was mainly interested in testing different positional enconding, thus I avoided hard coding them in the cache files so I don't have to recompute the cache everytime.

|> 3. atom_rrwp's docstring describes a different return than the code:

I inded ablated the pe_steps on PNA, and I included trails with 21 pe steps 


|> 4. pipe.py is unreachable

for now, I was planning to Implement a PE that include both Positional Enconding and Homology Enconding, not sure if it is worth right now.
